File: leetcode/c++/easy/_3151_special_array_1.cpp
// Leetcode Problem 3151

#include <gtest/gtest.h>

using namespace std;

class Solution {
public:
    bool isArraySpecial(vector<int>& nums) {
        if (nums.size() < 2) {
            return true;
        }

        for (int i = 0; i < nums.size() - 1; i++) {
            if ((nums[i] % 2) == (nums[i + 1] % 2)) {
                return false;
            }
        }
        return true;
    }
};

class TestIsArraySpecial : public ::testing::Test {
protected:
    Solution solution;
};

TEST_F(TestIsArraySpecial, SimpleTestCase1) {
    vector<int> nums = {1};

    EXPECT_EQ(solution.isArraySpecial(nums), true);
}

TEST_F(TestIsArraySpecial, SimpleTestCase2) {
    vector<int> nums = {2, 1, 4};

    EXPECT_EQ(solution.isArraySpecial(nums), true);
}

TEST_F(TestIsArraySpecial, SimpleTestCase2) {
    vector<int> nums = {4, 3, 1, 6};

    EXPECT_EQ(solution.isArraySpecial(nums), false);
}

File: leetcode/rust/problems/src/easy/_3151_special_array_1.rs
// Leetcode Problem 3151

struct Solution;

#[allow(dead_code)]
impl Solution {
    pub fn is_array_special(nums: Vec<i32>) -> bool {
        let mut last_parity = nums[0] % 2 == 0;

        for i in 1..nums.len() {
            if (nums[i] % 2 == 0) == last_parity {
                return false;
            }
            last_parity = !last_parity;
        }

        true
    }
}


#[test]
fn test_is_array_special() {
    let nums1 = vec![1];
    let nums2 = vec![2, 1, 4];
    let nums3 = vec![4, 3, 1, 6];

    assert_eq!(Solution::is_array_special(nums1), true);
    assert_eq!(Solution::is_array_special(nums2), true);
    assert_eq!(Solution::is_array_special(nums3), false);
}